boot2docker setup: make test fails in test TestInitializeCannotStatPathFileNameTooLong

Description of problem:

make test fails with 
--- FAIL: TestInitializeCannotStatPathFileNameTooLong (0.00s)
        volume_test.go:56: Expected to get ENAMETOOLONG error, got stat ivZgXsMAvLAYitpGtstNvMXsAaAVBmToJKPxJVufDWaiBQQslaOhPAoEuIjbVkMcmDgJgkfobtzfZolRdQoaheGbhjnxzkvihGxbmnSySnAbwimRPMFxrJcxuiwsQaQgChZshHzNRRAxQyfPepZcKWmcFjtXQoFCyMIPPsIMJXvKipMxgPXeROmvaabKYFjnJYVHzmliRDpvIHgazZEtFKXBBjWHZQbzsFFcXUvVYRolrzHDJqzmagfkxzxKWjGwshQrEbbVLsmHcgMxevSjaIIqigrpoahVQcrNLFssLHoQ: protocol error

Test environment:

On a mac (OS X 10.9.5) using boot2docker (1.5.0) and docker (1.5.0 - came with boot2docker). 
In the build docker container running in Virtualbox

`docker version`:

Client version: 1.5.0
Client API version: 1.17
Go version (client): go1.4.1
Git commit (client): a8a31ef
OS/Arch (client): darwin/amd64
Server version: 1.5.0
Server API version: 1.17
Go version (server): go1.4.1
Git commit (server): a8a31ef


`docker info`:

Containers: 2
Images: 46
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 50
Execution Driver: native-0.2
Kernel Version: 3.18.5-tinycore64
Operating System: Boot2Docker 1.5.0 (TCL 5.4); master : a66bce5 - Tue Feb 10 23:31:27 UTC 2015
CPUs: 8
Total Memory: 1.961 GiB
Name: boot2docker
ID: 3M76:UYVC:AQ6C:VPU4:L3LE:KJOM:QRBH:RLII:JCQO:7HV4:L3XD:7EDO
Debug mode (server): true
Debug mode (client): false
Fds: 20
Goroutines: 29
EventsListeners: 0
Init Path: /usr/local/bin/docker
Docker Root Dir: /mnt/sda1/var/lib/docker




`uname -a`:

Darwin xxxxxxxxx 13.4.0 Darwin Kernel Version 13.4.0: Wed Mar 18 16:20:14 PDT 2015; xxxxxxxx/RELEASE_X86_64 x86_64


Environment details (AWS, VirtualBox, physical, etc.):

VirtualBox (boot2docker): uname -a

Linux boot2docker 3.18.5-tinycore64 #1 SMP Sun Feb 1 06:02:30 UTC 2015 x86_64 GNU/Linux


How reproducible:

Every time


Steps to Reproduce:
1. Setup development environment on Mac OS X using boot2docker as per https://docs.docker.com/project/set-up-dev-env/
2. Start the docker container using (no changes yet in dry-run-test branch)
docker run --privileged --rm -ti -v `pwd`:/go/src/github.com/docker/docker dry-run-test /bin/bash
3. Run basic unit test using
hack/make.sh dynbinary binary cross test-unit


Actual Results:

+ go test -test.timeout=30m github.com/docker/docker/volumes
--- FAIL: TestInitializeCannotStatPathFileNameTooLong (0.00s)
        volume_test.go:53: Expected to get ENAMETOOLONG error, got stat EyCwAvvNYrBMvsDjqaaNMWKtmedHadMhXxIGiLNmuq\
jnMtJMNocxiicSnfzNJyscOJzzHHCybsUClNSkDSDtZVLKfuYuyokRuXfgZLIkVypeSfpAgxiTLWJDVyAJboesBvYYtWOXotYQqNcEXgfrVNGXPHJI\
jnyBRHjqScGvCbVuvynkwAXWubzjSSRitIwlktUbKSyBRuPaXsnZFmJaOHqMLxsVrHEFSAwUkpVxpdSvtPFBHTCvhIzmdahHxdertyaUJQGHtIkIBz\
UTCfdNGNqydbufaVhAospgpLgarxQi: protocol error
FAIL


Expected Results:

The test TestInitializeCannotStatPathFileNameTooLong expects that the os.Stat returns ErrNotExist
when given a long path name. Instead is returns "protocol error"

Additional info:

It appears os.Stat returns "protocol error" rather than ErrNotExist because something is different in the 
stat system call behavior in the following kernel that is running in VirtualBox VM.

Linux boot2docker 3.18.5-tinycore64 #1 SMP Sun Feb 1 06:02:30 UTC 2015 x86_64 GNU/Linux

So is this a boot2docker issue or docker issue - my guess is that it is docker issue in the test code itself - volumes/volume_test.go - since it is assuming stat will fail with only "file name too long" error.

--- a/volumes/volume_test.go
+++ b/volumes/volume_test.go
@@ -1,7 +1,7 @@
 package volumes
 
 import (
-	"strings"
+	"os"
 	"testing"
 
 	"github.com/docker/docker/pkg/stringutils"
@@ -33,8 +33,8 @@ func TestInitializeCannotMkdirOnNonExistentPath(t *testing.T) {
 		t.Fatal("Expected not to initialize volume with a non existent path")
 	}
 
-	if !strings.Contains(err.Error(), "mkdir : no such file or directory") {
-		t.Fatalf("Expected to get mkdir no such file or directory, got %s", err)
+	if !os.IsNotExist(err) {
+		t.Fatalf("Expected to get ErrNotExist error, got %s", err)
 	}
 }
 
@@ -49,7 +49,7 @@ func TestInitializeCannotStatPathFileNameTooLong(t *testing.T) {
 		t.Fatal("Expected not to initialize volume with a non existent path")
 	}
 
-	if !strings.Contains(err.Error(), "file name too long") {
-		t.Fatalf("Expected to get ENAMETOOLONG error, got %s", err)
+	if os.IsNotExist(err) {
+		t.Fatal("Expected to not get ErrNotExist")
 	}
 }

